aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -3384,5 +3387,69 @@ Tensor kron(const Tensor& self, const Tensor& other) {
   return KronImpl(self, other).kron();
 }
 
+// Weight Only Quantization Gemm
+DEFINE_DISPATCH(weight_to_int4pack_stub);
+DEFINE_DISPATCH(int4pack_mm_stub);
+
+Tensor _convert_weight_to_int4pack_cpu(
+    const Tensor& in,
+    int64_t innerKTiles) {
+
+  TORCH_CHECK(in.dim() == 2,
+      "_convert_weight_to_int4pack: expect weight to be 2D tensor.");
+  TORCH_CHECK(in.dtype() == at::kInt,
+      "_convert_weight_to_int4pack: expect weight to be kInt.");
+
+  auto weight = in.contiguous();
+  auto N = weight.size(0);
+  auto K = weight.size(1);
+
+  TORCH_CHECK(N % 16 == 0,
+      "_convert_weight_to_int4pack: expect N to be dividable by 16");
+  TORCH_CHECK(K % 2 == 0,
+      "_convert_weight_to_int4pack: expect K to be dividable by 2");
+
+  auto weight_packed = at::empty({N, K / 2}, weight.options().dtype(kByte));
+  weight_to_int4pack_stub(kCPU, weight_packed, weight);
+  return weight_packed;
+}
+
+Tensor _weight_int4pack_mm_cpu(
+    const Tensor& A,
+    const Tensor& B,
+    int64_t qGroupSize,
+    const Tensor& qScaleAndZeros) {
+
+  auto M = A.size(0);
+  auto N = B.size(0);
+  auto K = A.size(1);
+
+  TORCH_CHECK(A.dtype() == kBFloat16,
+      "_weight_int4pack_mm: expect A to be bfloat16 tensor.");
+  TORCH_CHECK(A.is_contiguous(),
+      "_weight_int4pack_mm: expect A to be contiguous.");
```

Comment:
Nit (use `__func__` rather than repeating function name)
```suggestion
      __func__, ": expect A to be contiguous.");
```

Reply:
updated